### DISCUSSION TITLE
Looking for Eglot pop-up menu like functionality:

### DISCUSSION THREAD

**Discussion by zeta-00** [joaotavora/eglot#912] [Ideas]
Greetings,

I just god Eglot working with Company, but, I was wondering if Eglot has equivalent functionality to LSP-UI, like a popup-menu, documentation, and so forth?

Or will I need enable some extra company feature to get this type of functionality with Eglot?

https://github.com/company-mode/company-quickhelp
^ This extra company-quickhelp package looks useful, but I want to check if Eglot has it's own built-in UI functionality to implement stuff like this?

Note: I'm transitioning from lsp-mode/lsp-ui to Eglot, so there's a lot I don't know and Eglot is still lacking in the documentation area. Eglot itself seems similar to lsp-mode, but I'm not sure what's the best way to get lsp-ui like functionality working with Eglot?

I'm confused about this, so I figured I'd ask here.

Thanks in advance for the help.

**Comment by trev-dev** [comment]
With eglot your choices are:

1. Eldoc

With this in mind you can look at packages that hook into Eldoc to "pop-up" information in places other than the mini-buffer.

There are two solutions that I have tried that work well.

1. https://repo.or.cz/eldoc-overlay.git - Eldoc Overlay. This one works well out of the box. Make sure to activate company before you activate eldoc-overlay. This is just as annoying as Eldoc, but it puts the obnoxious buffer-size changes next to your cursor instead :partying_face: 
2. https://github.com/casouri/eldoc-box - My personal favorite, though, it only works well if you configure it just so:
    ```elisp
    (setq eldoc-box-cleanup-interval 2
          eldoc-box-max-pixel-height 900
          eldoc-box-max-pixel-width 1500)
     ``` 
    
The `max-pixels` are up to your taste, but the defaults on most modern screens will cause lines to wrap obnoxiously early. 

The `eldoc-box-cleanup-interval` is gosh-darn mandatory. Without it, eldoc-box will interrupt your one and only thread if you move around too quickly and it m-m-m-makes things -s-t-st-stutter.  Keeping the box alive longer means no need to keep re-rendering the box.

Want to use the `eldoc-box-hover-at-point-mode`? Forget it! You may as well go back to LSP mode and `lsp-ui` and abandon any further hope for minimalism you had because the performance of this mode is abhorrent. Want to stop hiding the pop -over until you have enough curiosity to render one with `eldoc-box-eglot-help-at-point`? Forget it! It's broken.

---

**Edit:** Eldoc Box has gotten a lot better, as there are now working functions that call up the box on-demand. There are two separate commands for when you have Eglot, and when you do not:

```elisp
(setq eldoc-echo-area-use-multiline-p nil
      eldoc-idle-delay 0.75)

(require 'eldoc-box)
(global-set-key (kbd "C-c i") #'eldoc-box-help-at-point)
(define-key eglot-mode-map (kbd "C-c I") #'eldoc-box-eglot-help-at-point)
```

Hope this helps.

**Comment by krisbalintona** [comment]
@trev-dev Regarding the annoying line wrapping in the `eldoc-box` frame, the key function to look at is `eglot--format-markup` I think. The line wrapping is caused because the documentation is sent as markdown and `eglot` uses markdown (of the Github flavor) to render it. The line wrapping issue is caused by how `gfm-view-mode` renders horizontal rules, i.e. "---". 

I wrote the following workaround which replaces "---" with something else. In this case, I replaced it with a sequence of hyphen-like characters that should fill the width of the popup frame without causing wrapping:

```emacs-lisp
  (defun kb/eglot--format-markup (markup)
    "Format MARKUP according to LSP's spec."
    (pcase-let ((`(,string ,mode)
                 (if (stringp markup) (list markup 'gfm-view-mode)
                   (list (plist-get markup :value)
                         (pcase (plist-get markup :kind)
                           ("markdown" 'gfm-view-mode)
                           ("plaintext" 'text-mode)
                           (_ major-mode))))))
      (with-temp-buffer
        (setq-local markdown-fontify-code-blocks-natively t)

        ;; Replace the horizontal rule, which is three hyphens in the markup,
        ;; with X number of hyphens-like characters, with X being enough to
        ;; cover the width of `eldoc-box-max-pixel-width'. We can't simply
        ;; replace with more hyphens since `gfm-view-mode' renders any set of
        ;; three hyphens as a horizontal rule
        (setq string (string-replace "---"
                                     (make-string (floor (/ eldoc-box-max-pixel-width (window-font-width))) ?⎺)
                                     string))

        (insert string)
        (let ((inhibit-message t)
              (message-log-max nil))
          (ignore-errors (delay-mode-hooks (funcall mode))))
        (font-lock-ensure)
        (string-trim (buffer-string)))))
  (advice-add 'eglot--format-markup :override #'kb/eglot--format-markup)
```

I tried looking at `markdown-hr-strings` as well since it seems relevant but I don't know much about `markdown-mode` so I settled with the workaround above instead. There should be some edge cases between font pixel widths but this should work most of the time for fixed-pitch fonts.

**Comment by cxa** [comment]
The `hr` problem is due to `markdown-mode` render it as the width as the window body:

https://github.com/jrblevin/markdown-mode/blob/master/markdown-mode.el#L3515

It's uncontrollable. So I just replace the `---` to empty to workaround this problem:

```elisp
  (defun cxa/filter-args/eglot--format-markup (args)
    (mapcar (lambda (markup)
	      (if (plistp markup)
		  (let* ((value (plist-get markup :value))
			 (kind (plist-get markup :kind))
			 (_ (string= kind "markdown"))
			 (value (string-replace "---" "" value)))
		    (plist-put markup :value value))
		markup))
	    args))

(advice-add 'eglot--format-markup :filter-args 'cxa/filter-args/eglot--format-markup)
```

**Reply by ambihelical** [reply]
Seems that "\n---\n" works better, less vertical space, probably safer too.  Also had to change it a little for 28.1, plistp isn't defined. Thanks for this in any case.

@joaotavora Should eglot be doing this itself, at least as an optional filter?  Any sort of child frame will display multiple lines of hyphens for horizontal rulers without it.

**Reply by ambihelical** [reply]
There is also a similar problem in markdown-mode independent of eglot due to text scaling.  

https://github.com/jrblevin/markdown-mode/issues/618

**Comment by trev-dev** [comment]
FYI I updated my original response to suit my newer experience with eldoc-box. https://github.com/joaotavora/eglot/discussions/912#discussioncomment-2498996
